### PR TITLE
Fixes #30946: Use Rails default for JSON root node

### PR DESCRIPTION
**Pull request by ehelms**
…equest

Back when the application had a V1 and V2 API, the V2 API needed to
adhere to the designed format. This included not having a root node
in the JSON output for the object being returned. Given the use of
Rabl was not enforced across the ecosystem of plugins, this was
disabled per request for any request that did not go through the
Rabl rendering engine. The configuration that controls this is a
class variable and is thus not thread safe.

To enable the use of a multi-threaded Puma deployment, this setting
has to be set at application boot and not per request. The previous
implementation returned the setting back to it's default value of
true so that the V1 API was unaffected. This has potential consequences
across other areas of code or plugins that relied on the `to_json`
method on a model to perform further manipulations and were coded
to expect the root.

The alternative approach is to avoid setting this setting all together
and drop it's use. The expectation here would be that all API implementations
should be using Rabl templates properly to follow the API V2 spec
and any plugins relying on attempting to render through to_json
will need to fix this or potentiall encounter a change in their
API responses to have the root node.

### CODE REVIEW

**Review by ekohl — COMMENTED** [review]
There is also this:

https://github.com/theforeman/foreman/blob/b10e4c081eafe28e170b2b3de1be83a069c7fd08/config/initializers/wrap_parameters.rb#L10-L13

Then there's

https://github.com/theforeman/foreman_ansible/blob/321bd8453ab0ffab6818ea487cf21fba5f3e0be5/app/models/ansible_role.rb#L8
https://github.com/redhataccess/foreman-plugin/blob/9e6cb2b1b477e0f7971b47e23ea79c51a9065ce9/redhat-access/test/dummy/config/initializers/wrap_parameters.rb#L11-L14
https://github.com/redhataccess/foreman-plugin/blob/9e6cb2b1b477e0f7971b47e23ea79c51a9065ce9/redhat-access/app/models/redhat_access/telemetry_configuration.rb#L6

That's at least what I found when I quickly grepped in checkouts on my disk.

**Comment by ehelms** [comment]
At least for the instances were it's being set in a sub-class those are safe from affecting the global state. As to the others, the initializer dates back nearly 8 years and is one of those fun times where the comment and the code were at odds:

https://github.com/theforeman/foreman/commit/feacea35f07f362d9e2c694a83516bbc902321a0
```
# Disable root element in JSON by default.
ActiveSupport.on_load(:active_record) do
 self.include_root_in_json = true
end 
```

@jlsherrill pointed out to me Katello played this game as well (and dates back all the way to Katello's initial open sourcing when it was a stand alone app): https://github.com/Katello/katello/blob/master/config/initializers/active_record.rb

Given we are seeing an at-odds usage across the main application and the largest plugin in the ecosystem, I am leaning towards removing this from the ecosystem as far as I can tell the default value for this setting in Rails is `false` (my testing seems to confirm).

**Review by lzap — DISMISSED** [review]
Nice, but doesn't this change the current behaviour? Previously this was disabled for V2 endpoints only, granted not safely, however it was. Now, this is disabled also for non-V2 endpoints.

**Comment by ehelms** [comment]
What non V2 endpoints do we have? I was concerned about the unknown endpoints. I figured with v1 officially dropped as of Foreman 2 we should have only v2 endpoints.

**Comment by ehelms** [comment]
[test foreman]

**Comment by lzap** [comment]
> What non V2 endpoints do we have? I

By that I meant UI controllers. I wonder if this changes some behavior.

**Comment by ehelms** [comment]
> By that I meant UI controllers. I wonder if this changes some behavior.

Yea good point, that is one of the risks we may see as fallout. I'm hoping the tests will identify any of these scenarios :crossed_fingers:

**Comment by tbrisker** [comment]
Looks like at least one test failure is related.

**Comment by lzap** [comment]
Unless there are some voices, I am all for this. Whatever we do to make things more simple is for good. This bug in particular was a nasty one to track down!

**Comment by ekohl** [comment]
From a conceptual point I think this is a good idea. It makes things predictable and a controller should specify its own behavior if it differs. I don't think I can comment on the implementation details too much due to a lack of knowledge from my side. It is probably a good idea to tell this to plugin authors so they can check.

File: config/initializers/wrap_parameters.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@
   wrap_parameters :format => [:json]
 end
 
-# Enable root element in JSON by default.
+# Disable root element in JSON by default.
 ActiveSupport.on_load(:active_record) do
-  self.include_root_in_json = true
+  self.include_root_in_json = false
```
Comment (tbrisker) [inline config/initializers/wrap_parameters.rb]: This is the default in rails 6:
https://github.com/rails/rails/blob/6-0-stable/activerecord/lib/active_record/serialization.rb#L10

So i think we can just drop it from the initializer.

**Review by tbrisker — CHANGES_REQUESTED** [review]
Looks like at least one test relies on the old behaviour

**Comment by tbrisker** [comment]
[test foreman] not sure if related or not, but looks like integration tests timed out

**Comment by tbrisker** [comment]
Looks like tests are stalling forever with this change, and katello tests are failing with `NameError: uninitialized constant Api::V2::RegistrationController`... did you rebase on latest develop?

**Comment by tbrisker** [comment]
Integration tests are failing in many places with this change, aborted them (since the retries will eventually cause the task to timeout). not sure what's causing that though, from a very quick glance the ui seems to be functioning fine.

**Comment by tbrisker** [comment]
The integration tests are failing since the current user sent to the UI is no longer wrapped with `user:{}`. This line:

https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/components/UserDropdowns.js#L34
needs to change to:
```
  const userInfo = get(user, 'current_user');
```
Edit: Or simply (which would allow removing the lodash requirement):
```
  const userInfo = user.current_user;
```
And layout fixture should be adjusted accordingly

**Comment by tbrisker** [comment]
I think js test failures are not related to this change, as they are about the snapshots for `HostDetailPage` missing some attributes, guessing a new patternfly version was released that added them? @amirfefer maybe you know?

**Comment by ekohl** [comment]
The commit message should probably be modified that we start to rely on the Rails default without providing a way to override it.

**Comment by tbrisker** [comment]
integration tests are still failing, aborted them after 1.5 hours before they time out

**Comment by ehelms** [comment]
I found the rest of the code that was relying on subnet object in the UI to have a root node.

**Review by tbrisker — APPROVED** [review]
Thanks @ehelms ! Can you please update on dev forum so that plugin maintainers are aware of this change in case they rely on the wrapped behaviour?

**Pull request merged**